src/psyclone/psyir/nodes/reference.py

```diff
@@ -257,6 +257,37 @@ def next_accesses(self):
         chain = DefinitionUseChain(self)
         return chain.find_forward_accesses()
 
+    def escapes_scope(self, scope, visited_nodes=None) -> bool:
```

Comment:
I also add @LonelyCat124 as a reviewer, particularly because I added this method that uses DefUseChain to decide if a reference can be private (by checking if it is ever used after a given scope, .e.g. a loop). I am also planning a symmetric method looping backwards to replace the logic about when the private should be firstprivate.